trinity/protocol/bcc/validators.py

```diff
@@ -0,0 +1,86 @@
+from typing import (
+    Tuple,
+    Union,
+)
+
+from eth_typing import (
+    Hash32,
+)
+
+from eth_utils import (
+    ValidationError,
+    encode_hex,
+)
+from eth_utils.toolz import (
+    sliding_window,
+)
+
+from eth.beacon.types.blocks import BaseBeaconBlock
+
+from trinity.protocol.common.validators import BaseValidator
+
+from trinity.protocol.bcc.commands import (
+    RequestMessage,
+    ResponseMessage,
+)
+
+
+class BeaconBlocksValidator(BaseValidator[Tuple[BaseBeaconBlock, ...]]):
+
+    def __init__(self, block_slot_or_hash: Union[int, Hash32], max_blocks: int) -> None:
+        self.block_slot_or_hash = block_slot_or_hash
+        self.max_blocks = max_blocks
+
+    def validate_result(self, result: Tuple[BaseBeaconBlock, ...]) -> None:
+        self._validate_first_block(result)
+        self._validate_number(result)
+        self._validate_sequence(result)
+
+    @property
+    def _is_numbered(self) -> bool:
+        return isinstance(self.block_slot_or_hash, int)
+
+    def _validate_first_block(self, blocks: Tuple[BaseBeaconBlock, ...]) -> None:
+        try:
+            first_block = blocks[0]
+        except IndexError:
+            return
+
+        if self._is_numbered:
+            if first_block.slot != self.block_slot_or_hash:
+                raise ValidationError(
+                    f"Requested blocks starting with slot #{self.block_slot_or_hash} but first "
+                    f"returned block is from slot #{first_block.slot}"
+                )
+        else:
+            if first_block.hash != self.block_slot_or_hash:
+                raise ValidationError(
+                    f"Requested blocks starting with hash {encode_hex(self.block_slot_or_hash)} "
+                    f"but first returned block has hash {encode_hex(first_block.hash)}"
+                )
+
+    def _validate_number(self, blocks: Tuple[BaseBeaconBlock, ...]) -> None:
+        if len(blocks) > self.max_blocks:
+            raise ValidationError(
+                f"Requested up to {self.max_blocks} blocks but received {len(blocks)}"
+            )
+
+    def _validate_sequence(self, blocks: Tuple[BaseBeaconBlock, ...]) -> None:
+        # workaround for https://github.com/pytoolz/cytoolz/issues/123#issuecomment-432905716
+        if not blocks:
+            return
+
+        for parent, child in sliding_window(2, blocks):
+            if child.parent_root != parent.hash:
```

Comment:
This appears to be different from the `eth` protocol, requiring that responses be contiguous (no gaps allowed).  Is that accurate?  Might be nice to add a comment indicating what exactly is being validated at each of these steps in human readable terms like we do here: https://github.com/ethereum/py-evm/blob/master/trinity/protocol/common/validators.py#L130

Reply:
Yes, it's because we don't have `reverse` and `skip`, so we can be a bit more strict here. I've added some comments.